src/Mvc/Mvc.Core/src/ModelBinding/Metadata/ModelAttributes.cs

```diff
@@ -214,6 +214,17 @@ public static ModelAttributes GetAttributesForParameter(ParameterInfo parameterI
 
     private static Type? GetMetadataType(Type type)
     {
-        return type.GetCustomAttribute<ModelMetadataTypeAttribute>()?.MetadataType;
+        // GetCustomAttribute will examine the members inheritance chain
+        // for attributes of a particular type by default. Meaning that
+        // in the following scenario, the `ModelMetadataType` attribute on
+        // both the derived _and_ base class will be returned.
+        // [ModelMetadataType<BaseModel>]
+        // private class BaseViewModel { }
+        // [ModelMetadataType<DerivedModel>]
+        // private class DerivedViewModel : BaseViewModel { }
+        // To avoid this, we call `GetCustomAttributes` directly
+        // to avoid examining the inheritance hierarchy.
+        // See https://source.dot.net/#System.Private.CoreLib/src/System/Attribute.CoreCLR.cs,677
+        return type.GetCustomAttributes<ModelMetadataTypeAttribute>(inherit: false).SingleOrDefault()?.MetadataType;
```

Comment:
If GetCustomAttributes returns multiple, throw an InvalidOperationException with a message that only one attribute of this type is allowed of the specified type.